### ISSUE TITLE
Not really an issue but...

### ISSUE THREAD

**Issue by rdbeach** [lovasoa/whitebophir#31]
Hey. Thanks for a great board. I have forked this board and made some alterations (actually quite a bit), mostly for my own personal use as a math tutor. Check them out at [rdbeach/wb](https://github.com/rdbeach/wb) So far, the changes have undergone limited testing on Safari and Chrome, both on my laptop and on my phone. Seems to work pretty good so far.

A list of changes:

Redesign of the GUI/layout
Added whiteout tool
Added circle tool
Modified text tool so that the input appears closer to the text
Made the removal tool more robust so you can remove small objects easier 
Added an optional grid
Added an eraser tool
Added a clear board tool
Added an undo tool
Added a redo tool
Added a zoom in and zoom out tool in separate files.
Made some changes to the server to accommodate the new tools.

**Comment by lovasoa** [comment]
Your work is great ! Congratulations @rdbeach !

I am reposting here [the comment I made on your commit](https://github.com/rdbeach/wb/commit/b532b7b8fec9348f7373430882fc5b8138833ed3#commitcomment-38503451) :

> Hello @rdbeach !
> I stumbled on this commit on my github feed. It looks like you've done some great work on WBO ! 👍 Would you be interested by working with me on integrating some of these changes back into WBO ? This would give them more visibility, reduce the maintenance burden on your fork, and potentially enable some more contributions by other parties. I am especially interested by your new eraser tool. Working with svg masks is a great idea that I wish I had from the beginning !
There are some browser compatibility and other issues with your changes, though, and I would be glad to work with you on solving these.
> Cheers,
> @lovasoa  

Also, I think you forgot to commit some files. I am not seeing the eraser tool, and the undo/redo don't seem to work...
It would be awesome if we could work together on polishing and integrating these changes.

<img width="1171" alt="image" src="https://user-images.githubusercontent.com/552629/79316966-638d7380-7f05-11ea-8670-ee80b59cb517.png">

**Comment by rdbeach** [comment]
Hi. Sure, just let me know what you need me to do. Obviously, time allowing.

Here is a link to a running demo: https://lcfprep.com/wb2/wb.html

Probably some kinks to work out, since I've only been working on it on my
system.  I changed the server a bit also, btw.

The eraser tool works, but there is really no way to make a really good
eraser with svg. It tends to get sluggish quickly and is best limited to
"touch ups". That's why I hid it in the popout tool. So people would get
used to using the removal tool instead.

**Comment by lovasoa** [comment]
There is currently a single big commit. It would be good to make several small pull requests that we can discuss and work on independently. 
Some of them can already be merged now, such as 

 - [ ] GUI/layout improvements
 - [ ] the circle tool
 - [ ] Modified text tool so that the input appears closer to the text
 - [ ] Made the removal tool more robust so you can remove small objects easier
 - [ ] Added an optional grid

**Comment by rdbeach** [comment]
How do you break a pull request in smaller ones?

Maybe make a branch for different features? I can fill out the minimal code changes for each branch and we can review and discuss. But maybe just one branch at a time to keep things simpler.

**Comment by lovasoa** [comment]
Yes, you are right, doing the PRs one by one will avoid conflicts and keep things simpler. Just create a branch with the changes and open a PR from that branch to master on lovasoa/whitebophir. Maybe let's start with the smalleest and least controversial things, like the circle tool or the change in the position of the text input field.

**Comment by rdbeach** [comment]
Hi. Quick update. I added a transform tool, and an area eraser. I'm actually using the board right now with some of my clients, and I will probably do some more rigorous testing of it in the days to come. I've been doing this partly for my work, and partly as a hobby to keep me busy during this difficult time. Unfortunately, I've made so many changes to the original board that I don't know if its feasible (or worth the effort) to merge them at this point. One more feature I may want to incorporate is loading a background image.

**Comment by lovasoa** [comment]
A background image tool was implemented in https://github.com/lovasoa/whitebophir/pull/21 

I will look into merging your work as it is tested. Do you plan to support firefox ?

**Comment by rdbeach** [comment]
I just downloaded the latest Firefox a few hours ago. At first, it looked like it wasn't working at all, but it turned out to be just a few minor glitches, and now it looks good, as far as I can tell. Should I support edge as well?  Haven't tried that one.

**Comment by finnboeger** [comment]
Hi I had a go at starting to split the changes into smaller more manageable chunks.
For now I have split some inital changes to the board initialization and the server socket handling off into [base-changes](https://github.com/finnboeger/whitebophir/tree/base-changes) and the gui changes into [style-changes](https://github.com/finnboeger/whitebophir/tree/style-changes). 
The code for the improved removal tool has been split off to [tool/eraser](https://github.com/finnboeger/whitebophir/tree/tool/eraser), however it - like the future branches I will create - depends on the first two changesets.

The next steps that I can see are
- [ ] The circle tool
- [ ] The transform tool
- [ ] The image tool
- [ ] The text changes
- [ ] The touch-up eraser, which requires adding support for layers to every tool
- [ ] The cursor - this massively increases the amount of messages being sent, should possibly be throttled
- [ ] The grid
- [ ] The separate zoom buttons
- [ ] The undo, redo and clear tools which require a server side undo/redo history and new message type. Also add a confirmation dialog to the clear button and possibly restrict it
- [ ] > A UX improvement in conjunction with the 3 above: implement them as simple buttons instead of tools, reducing the amount of clicks necessary to e.g. toggle the grid

**Comment by lovasoa** [comment]
Great, let's discuss that in https://github.com/lovasoa/whitebophir/pull/38

**Comment by rdbeach** [comment]
"The touch-up eraser, which requires adding support for layers to every tool"

Also, the transform tool requires each drawing element to set their transform property.

**Comment by rdbeach** [comment]
BTW has anyone discussed client configuration? I see that you updated the backend configuration. Probably when you start putting in so many features, not everyone wants or needs them. Maybe another config file for the front end?

**Comment by finnboeger** [comment]
I realized that part about the transform tool when I had a closer look at it, thats why I'm holding off on implementing it until the other tools have been merged.

As for front end configuration I have added your latest changes to the cursor to allow disabling it, but there is no settings page yet. I believe adding a `oneTouch` tool that would open a setting modal might be the best way to go?

I would also like to have server and board wide settings to allow you to disable certain features for all boards or on board creation, e.g. you might not want the document upload feature on a public server or the cursors on a low power server or very high use board.

**Comment by lovasoa** [comment]
Yes, having the configuration reflected on the client is probably a good idea. It could be done the same way as it is [for translations](https://github.com/lovasoa/whitebophir/blob/master/client-data/board.html#L80).

Also, what would you think about switching to using modern javascript, and a transpiler ? This would make the client-side code more readable in many instances.

**Comment by finnboeger** [comment]
I definitely see the benefit however I have no experience setting it up :laughing: 
We use typescript at work but I believe there would be quite a bit of effort in getting everything ready for such a switch.

**Comment by lovasoa** [comment]
I can set it up. It will just take some time, and you will probably have to rebase your PRs. Let's finish the "easy" PRs, and then I'll get to work. What we can merge without much changes are your current PRs for : 
 - circle tool
 - improved text tool
 - cursor

I added comments on those.

**Comment by finnboeger** [comment]
Another question for my testing: Which browsers should we support? 
I am currently only testing in Chrome and Firefox as I am on Linux and don't have any macOS or iOS devices to test on them. 
I can perform tests in Internet Explorer if necessary but I would much rather let it die. As the new Edge is based on Chromium as well I believe it should automatically be supported.

**Comment by finnboeger** [comment]
When integrating all the currently proposed features I'd recommend a new Menu structure:
![wbo menu](https://user-images.githubusercontent.com/10922421/80424703-8b57e080-88e2-11ea-873d-ff532b674bf5.png)
The icons are obviously not final and we'd have to rethink how we handle hovers (maybe ignore hover over the tooltip and only attach the hover event to the icon itself).
It is also currently missing the touch up eraser as I'm not sure where to put it without creating too much confusion or incentivising its use too much

What do you think?

**Comment by rdbeach** [comment]
Right now I am testing on Chrome, Safari, and Firefox. Recent versions only. Also I have tested it on safari and chrome for iPhone. Have not tested android phones. I will most likely continue to test daily, especially since I am using it with my clients.

**Comment by lovasoa** [comment]
@finnboeger : 👍  to the two-columns menu on desktop. We'll have to see if it doesn't take too much space on small screens.

@rdbeach : WBO was started a long time ago, and it was initially designed to work on internet explorer 9+ too. However, I don't test it on IE on a regular basis...

**Comment by finnboeger** [comment]
Regardless whether we make the change to the two-column layout we should offset the canvas to the right by the width of the menu. Otherwise if you e.g. write something and then zoom out it gets occluded by the menu.

**Comment by lovasoa** [comment]
What about adding a button to hide the menu instead ? On small screens, I think we cannot afford to reduce the size of the canvas...

**Comment by finnboeger** [comment]
Very good point. How about on small devices we hide the menu by default and only display the currently active tool and maybe the size indicator (using the old UI where it showed a dot of the current size but with the current color and opacity)? 
When clicking on the tool it would open up an overlay with the available tools whereas clicking on the size indicator would bring up a combined menu for changing size, color and opacity.

Tying into that I noticed another UX issue in my last lecture: 
New Users don't know about the available functionality. I believe adding a screen explaining all available features for users that are opening it for the first time would be helpful. That could also explain how to select their tool for mobile users.
I have created a couple of mockups:

Mobile UI:
![mobileUI](https://user-images.githubusercontent.com/10922421/80608660-61fc9900-8a37-11ea-8e92-1449fa830551.png)

Mobile UI hidden:
![mobileUIhidden](https://user-images.githubusercontent.com/10922421/80608719-79d41d00-8a37-11ea-8d69-9ca1dc95a398.png)

Mobile UI expanded:
![mobileUIopen](https://user-images.githubusercontent.com/10922421/80610375-95402780-8a39-11ea-9b4f-a3340df49022.png)

Intro / Help Screen:
![wbo intro](https://user-images.githubusercontent.com/10922421/80624884-9cbcfc00-8a4c-11ea-8ee8-ce7ce6ff1b97.png)

I added a whole load of features in the mockups but I don't expect that we will be able to implement them all directly.

What do you think?

**Comment by tomichGIT** [comment]
hi guys, thanks for this amazing script. I have a question do... if you are working together, should we be using @lovasoa or @rdbeach github repo? i loved the changes made by @rdbeach for teaching math.. but i dont want to stick with an unopdated version. Are you guys merging into one? thanks a lot again!

**Comment by lovasoa** [comment]
Hello! Yes, a lot of the changes are being merged, have a look at the pull requests! @finnboeger is doing a fantastic job at splitting the features into distinct PRs, fixing bugs, making the code cleaner and more maintainable. This represents a decent amount of work, so if you want to contribute on the existing pull requests, please do!

**Comment by rdbeach** [comment]
Hey, what's going on over here? I see you've made some changes. Where did @finnboeger go? I'm almost done with the buildout now. Maybe we can work together to clean up the code a bit. If you want to take it to the next level, we should make an organizational account and put the repo in there, where we can work as collaborators. It would probably attract more interest as well, because, then, it's not like you're just contributing to someone's pet project.

**Comment by lovasoa** [comment]
Hello ! Good that you are almost done with your changes, we'll be able to integrate more of them ! I think there is quite some work, so feel free to open new PRs and comment on the ones of @finnboeger. Ideally, we'll keep one feature by PR, to keep things manageable. We can create an organization, but if you're okay with it, I'd still like to review each change individually, and avoid giving external commit access immediately.

**Comment by mwllgr** [comment]
Hi, any news on these changes? The grid and dashed functions would be really great to have. @lovasoa @rdbeach

**Comment by lovasoa** [comment]
This issue has become too broad to keep as an actionable tracker.

Many items from the original fork discussion have since either landed or have their own focused issues:

- circle/ellipse-style shape support is present in the current tool set;
- grid, clear, zoom, cursor/presence, image/background-related work, and many toolbar/runtime changes have been integrated over time;
- clear is now a capability-gated tool rather than an always-on destructive button;
- user/presence work is tracked separately and now has concrete implementation;
- undo/redo remains open as #110;
- canvas/background color remains open as #96;
- help/shortcut discoverability remains open as #87;
- mobile/toolbar ergonomics remain open in focused issues such as #231.

The remaining useful work should be tracked in those narrower issues or in new focused issues/PRs. Keeping this 2020 umbrella open makes triage harder because it mixes already-implemented tools, rejected implementation details, and unrelated future UI proposals.

Closing this as superseded by the focused issues above.
